// _roadhog@2.2.0@roadhog/lib/utils/mock.js
"use strict";

var _interopRequireWildcard = require("@babel/runtime/helpers/interopRequireWildcard");

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.applyMock = applyMock;
exports.outputError = outputError;

var _keys = _interopRequireDefault(require("@babel/runtime/core-js/object/keys"));

var _fs = _interopRequireWildcard(require("fs"));

var _assert = _interopRequireDefault(require("assert"));

var _chokidar = _interopRequireDefault(require("chokidar"));

var _chalk = _interopRequireDefault(require("chalk"));

var _expressHttpProxy = _interopRequireDefault(require("express-http-proxy"));

var _url = _interopRequireDefault(require("url"));

var _path = require("path");

var _bodyParser = _interopRequireDefault(require("body-parser"));

var _getPaths = _interopRequireDefault(require("../getPaths"));

var _winPath = _interopRequireDefault(require("./winPath"));

var debug = require('debug')('roadhog:mock');

var error = null;
var paths = (0, _getPaths.default)(process.cwd());
var configFile = paths.resolveApp('.roadhogrc.mock.js');
var mockDir = paths.resolveApp('./mock/');

function getConfig() {
  if ((0, _fs.existsSync)(configFile)) {
    // disable require cache
    (0, _keys.default)(require.cache).forEach(function (file) {
      if (file === configFile || file.indexOf(mockDir) > -1) {
        debug("delete cache ".concat(file));
        delete require.cache[file];
      }
    });
    return require(configFile);
  } else {
    return {};
  }
}

function createMockHandler(method, path, value) {
  return function mockHandler() {
    var res = arguments.length <= 1 ? undefined : arguments[1];

    if (typeof value === 'function') {
      value.apply(void 0, arguments);
    } else {
      res.json(value);
    }
  };
}

function createProxy(method, path, target) {
  return (0, _expressHttpProxy.default)(target, {
    filter: function filter(req) {
      return method ? req.method.toLowerCase() === method.toLowerCase() : true;
    },
    forwardPath: function forwardPath(req) {
      var matchPath = req.originalUrl;
      var matches = matchPath.match(path);

      if (matches.length > 1) {
        matchPath = matches[1];
      }

      return (0, _winPath.default)((0, _path.join)(_url.default.parse(target).path, matchPath));
    }
  });
}

function applyMock(devServer) {
  try {
    realApplyMock(devServer);
    error = null;
  } catch (e) {
    console.log(e);
    error = e;
    console.log();
    outputError();

    var watcher = _chokidar.default.watch([configFile, mockDir], {
      ignored: /node_modules/,
      ignoreInitial: true
    });

    watcher.on('change', function (path) {
      console.log(_chalk.default.green('CHANGED'), path.replace(paths.appDirectory, '.'));
      watcher.close();
      applyMock(devServer);
    });
  }
}

function realApplyMock(devServer) {
  var config = getConfig();
  var app = devServer.app;
  devServer.use(_bodyParser.default.json({
    limit: '5mb'
  }));
  devServer.use(_bodyParser.default.urlencoded({
    extended: true,
    limit: '5mb'
  }));
  (0, _keys.default)(config).forEach(function (key) {
    var keyParsed = parseKey(key);
    (0, _assert.default)(!!app[keyParsed.method], "method of ".concat(key, " is not valid"));
    (0, _assert.default)(typeof config[key] === 'function' || typeof config[key] === 'object' || typeof config[key] === 'string', "mock value of ".concat(key, " should be function or object or string, but got ").concat(typeof config[key]));

    if (typeof config[key] === 'string') {
      var path = keyParsed.path;

      if (/\(.+\)/.test(path)) {
        path = new RegExp("^".concat(path, "$"));
      }

      app.use(path, createProxy(keyParsed.method, path, config[key]));
    } else {
      app[keyParsed.method](keyParsed.path, createMockHandler(keyParsed.method, keyParsed.path, config[key]));
    }
  }); // 调整 stack，把 historyApiFallback 放到最后

  var lastIndex = null;

  app._router.stack.forEach(function (item, index) {
    if (item.name === 'webpackDevMiddleware') {
      lastIndex = index;
    }
  });

  var mockAPILength = app._router.stack.length - 1 - lastIndex;

  if (lastIndex && lastIndex > 0) {
    var newStack = app._router.stack;
    newStack.push(newStack[lastIndex - 1]);
    newStack.push(newStack[lastIndex]);
    newStack.splice(lastIndex - 1, 2);
    app._router.stack = newStack;
  }

  var watcher = _chokidar.default.watch([configFile, mockDir], {
    ignored: /node_modules/,
    persistent: true
  });

  watcher.on('change', function (path) {
    console.log(_chalk.default.green('CHANGED'), path.replace(paths.appDirectory, '.'));
    watcher.close(); // 删除旧的 mock api

    app._router.stack.splice(lastIndex - 1, mockAPILength);

    applyMock(devServer);
  });
}

function parseKey(key) {
  var method = 'get';
  var path = key;

  if (key.indexOf(' ') > -1) {
    var splited = key.split(' ');
    method = splited[0].toLowerCase();
    path = splited[1];
  }

  return {
    method: method,
    path: path
  };
}

function outputError() {
  if (!error) return;
  var filePath = error.message.split(': ')[0];
  var relativeFilePath = filePath.replace(paths.appDirectory, '.');
  var errors = error.stack.split('\n').filter(function (line) {
    return line.trim().indexOf('at ') !== 0;
  }).map(function (line) {
    return line.replace("".concat(filePath, ": "), '');
  });
  errors.splice(1, 0, ['']);
  console.log(_chalk.default.red('Failed to parse mock config.'));
  console.log();
  console.log("Error in ".concat(relativeFilePath));
  console.log(errors.join('\n'));
  console.log();
}
