// _roadhog@2.2.0@roadhog/lib/getWebpackConfig.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault");

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = _default;

var _toConsumableArray2 = _interopRequireDefault(require("@babel/runtime/helpers/toConsumableArray"));

var _extends2 = _interopRequireDefault(require("@babel/runtime/helpers/extends"));

var _getConfig = _interopRequireDefault(require("af-webpack/getConfig"));

var _browsers = _interopRequireDefault(require("./defaultConfigs/browsers"));

var _getEntry = _interopRequireDefault(require("./utils/getEntry"));

var debug = require('debug')('roadhog:getWebpackConfig');

var isDev = process.env.NODE_ENV === 'development';

function _default() {
  var opts = arguments.length > 0 && arguments[0] !== undefined ? arguments[0] : {};
  var cwd = opts.cwd,
      config = opts.config,
      babel = opts.babel,
      paths = opts.paths;
  var browserslist = config.browserslist || _browsers.default;
  debug("babel: ".concat(babel));
  debug("browserslist: ".concat(browserslist));
  return (0, _getConfig.default)((0, _extends2.default)({
    cwd: cwd
  }, config, {
    entry: (0, _getEntry.default)({
      cwd: paths.appDirectory,
      entry: config.entry,
      isBuild: !isDev
    }),
    babel: config.babel || {
      presets: [[babel, {
        browsers: browserslist
      }]].concat((0, _toConsumableArray2.default)(config.extraBabelPresets || [])),
      plugins: config.extraBabelPlugins || []
    },
    browserslist: browserslist
  }));
}
